File: app/src/main/java/com/kml/aGlobalUses/ExternalDbHelper.java
package com.kml.aGlobalUses;

import android.util.Log;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

public abstract class ExternalDbHelper extends Thread
{
    public static final String IO_EXCEPTION_TAG = "IO_EXCEPTION_TAG";

    public HttpURLConnection setConnection(String address)
    {
        HttpURLConnection conn = null;
        try {
            URL url = new URL(address);
            conn = (HttpURLConnection) url.openConnection();
            conn.setDoOutput(true);
            conn.setDoInput(true);
            conn.setRequestMethod("POST");

        }
        catch (IOException e)
        {
            Log.d(IO_EXCEPTION_TAG, "IOException setConnection: "+e.getMessage());
        }
        return conn;
    }

    public String readResult(HttpURLConnection conn)
    {
        InputStream inputStream;
        StringBuilder readResult = new StringBuilder();
        try {
            inputStream = conn.getInputStream();
            BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream,"UTF-8"));
            String line;

            while((line=reader.readLine()) != null)
            {
                readResult.append(line);
            }
            inputStream.close(); reader.close();
            conn.disconnect();
        } catch (IOException e) {
            Log.d(IO_EXCEPTION_TAG, "readResult: "+e.getMessage());
        }

        return readResult.toString();
    }

}

File: app/src/main/java/com/kml/data/networking/ExternalDbHelper.kt
package com.kml.data.networking

import android.os.Handler
import android.os.Looper
import com.kml.data.listeners.OnResultListener
import com.kml.extensions.logError
import com.kml.extensions.logNetworkResponse
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

abstract class ExternalDbHelper : Thread() {

    lateinit var listener: OnResultListener

    companion object {
        const val IO_EXCEPTION_TAG = "IO_EXCEPTION_TAG"
        const val BASE_URL = "http://192.168.1.111/KmlApi2/" //"http://sobos.ssd-linuxpl.com/KmlApi3/"
    }

    protected fun setConnection(address: String): HttpURLConnection? {
        var conn: HttpURLConnection? = null

        try {
            val url = URL(address)
            conn = url.openConnection() as HttpURLConnection
            conn.doOutput = true
            conn.doInput = true
            conn.requestMethod = "POST"
        } catch (e: Exception) {
            logError(e)
        }

        return conn
    }

    protected fun readResult(conn: HttpURLConnection?): String {
        var readResult = ""
        try {
            val inputStream = conn?.inputStream
            val reader = BufferedReader(InputStreamReader(inputStream, "UTF-8"))
            var line: String?

            while (reader.readLine().also { line = it } != null) {
                readResult += line
            }

            inputStream?.close(); reader.close(); conn?.disconnect()
        } catch (e: Exception) {
            logError(e)
        }
            logNetworkResponse(readResult)
        return readResult
    }

    protected fun invokeOnReceive(result: String) {
        Handler(Looper.getMainLooper()).post { listener.onReceive(result) }
    }

    fun setOnResultListener(operation: (String) -> Unit) {
        this.listener = object : OnResultListener {
            override fun onReceive(result: String) {
                operation(result)
            }
        }
    }
}